File: android/app/src/main/kotlin/KokoroNative.kt
package com.example.nova_voice_assistant

import android.content.Context
import ai.onnxruntime.*
import java.nio.FloatBuffer
import java.nio.ByteBuffer
import java.nio.ByteOrder

/**
 * Native wrapper for Kokoro TTS ONNX model
 * Handles text-to-speech synthesis using ONNX Runtime
 */
class KokoroNative(private val context: Context) {
    private var ortEnvironment: OrtEnvironment? = null
    private var ortSession: OrtSession? = null
    private var isModelLoaded = false
    
    // TTS configuration
    private val SAMPLE_RATE = 16000
    private val HOP_LENGTH = 256
    
    companion object {
        init {
            System.loadLibrary("onnxruntime")
        }
    }
    
    /**
     * Load the Kokoro TTS ONNX model
     */
    fun loadModel(modelPath: String): Boolean {
        try {
            // Initialize ONNX Runtime environment
            ortEnvironment = OrtEnvironment.getEnvironment()
            
            // Create session options
            val sessionOptions = OrtSession.SessionOptions()
            sessionOptions.setIntraOpNumThreads(4)
            sessionOptions.setOptimizationLevel(OrtSession.SessionOptions.OptLevel.ALL_OPT)
            
            // Load model from assets or file path
            val modelBytes = if (modelPath.startsWith("assets/")) {
                context.assets.open(modelPath.removePrefix("assets/")).readBytes()
            } else {
                java.io.File(modelPath).readBytes()
            }
            
            // Create ONNX session
            ortSession = ortEnvironment?.createSession(modelBytes, sessionOptions)
            
            isModelLoaded = true
            return true
        } catch (e: Exception) {
            e.printStackTrace()
            return false
        }
    }
    
    /**
     * Synthesize speech from text
     * Returns PCM audio data (16-bit, 16kHz, mono)
     */
    fun synthesize(
        text: String,
        speed: Float = 1.0f,
        voice: String = "female"
    ): ByteArray {
        if (!isModelLoaded || ortSession == null) {
            throw IllegalStateException("Model not loaded")
        }
        
        try {
            // Prepare text input
            val preparedText = prepareText(text)
            
            // Convert text to token IDs (phonemes)
            val tokenIds = textToTokens(preparedText)
            
            // Create input tensors
            val inputShape = longArrayOf(1, tokenIds.size.toLong())
            val inputTensor = OnnxTensor.createTensor(
                ortEnvironment,
                tokenIds,
                inputShape
            )
            
            // Additional inputs for speaker embedding (for voice selection)
            val speakerEmbedding = getSpeakerEmbedding(voice)
            val speakerTensor = OnnxTensor.createTensor(
                ortEnvironment,
                FloatBuffer.wrap(speakerEmbedding),
                longArrayOf(1, speakerEmbedding.size.toLong())
            )
            
            // Run inference
            val inputs = mapOf(
                "input_ids" to inputTensor,
                "speaker_embedding" to speakerTensor
            )
            val outputs = ortSession!!.run(inputs)
            
            // Get mel-spectrogram output
            val melOutput = outputs[0].value as Array<Array<FloatArray>>
            val mel = melOutput[0]
            
            // Convert mel-spectrogram to audio waveform using Griffin-Lim or vocoder
            val audioFloats = melToWaveform(mel, speed)
            
            // Convert float audio to PCM 16-bit
            val pcmData = floatArrayToPCM(audioFloats)
            
            // Clean up
            inputTensor.close()
            speakerTensor.close()
            outputs.close()
            
            return pcmData
        } catch (e: Exception) {
            e.printStackTrace()
            throw RuntimeException("Speech synthesis failed: ${e.message}")
        }
    }
    
    /**
     * Prepare text for TTS (normalization, etc.)
     */
    private fun prepareText(text: String): String {
        return text
            .trim()
            .replace(Regex("\\s+"), " ")
            .lowercase()
    }
    
    /**
     * Convert text to phoneme token IDs
     * This is a simplified version - actual implementation needs a proper phonemizer
     */
    private fun textToTokens(text: String): LongArray {
        // TODO: Implement proper text-to-phoneme conversion
        // For now, simple character-to-token mapping
        val vocab = loadVocabulary()
        
        val tokens = mutableListOf<Long>()
        for (char in text) {
            val token = vocab[char.toString()] ?: 0L
            tokens.add(token)
        }
        
        return tokens.toLongArray()
    }
    
    /**
     * Load vocabulary/phoneme mapping
     */
    private fun loadVocabulary(): Map<String, Long> {
        // TODO: Load actual vocabulary from tokenizer config
        return emptyMap()
    }
    
    /**
     * Get speaker embedding for voice selection
     */
    private fun getSpeakerEmbedding(voice: String): FloatArray {
        // Return pre-computed speaker embeddings
        // These would be loaded from the model config
        return when (voice) {
            "male" -> FloatArray(256) { 0.5f }  // Placeholder
            else -> FloatArray(256) { 0.0f }    // Female (default)
        }
    }
    
    /**
     * Convert mel-spectrogram to audio waveform
     * Using Griffin-Lim algorithm or neural vocoder
     */
    private fun melToWaveform(mel: Array<FloatArray>, speed: Float): FloatArray {
        // This is a simplified placeholder
        // Actual implementation needs Griffin-Lim or a neural vocoder like HiFi-GAN
        
        val numFrames = mel.size
        val audioLength = (numFrames * HOP_LENGTH / speed).toInt()
        val audio = FloatArray(audioLength)
        
        // TODO: Implement actual mel-to-waveform conversion
        // For now, generate silence
        return audio
    }
    
    /**
     * Convert float audio [-1, 1] to PCM 16-bit
     */
    private fun floatArrayToPCM(audioFloats: FloatArray): ByteArray {
        val pcmData = ByteArray(audioFloats.size * 2)
        val buffer = ByteBuffer.wrap(pcmData).order(ByteOrder.LITTLE_ENDIAN)
        
        for (sample in audioFloats) {
            // Clamp to [-1, 1] and convert to 16-bit PCM
            val clampedSample = sample.coerceIn(-1.0f, 1.0f)
            val pcmSample = (clampedSample * 32767).toInt().toShort()
            buffer.putShort(pcmSample)
        }
        
        return pcmData
    }
    
    /**
     * Synthesize speech in chunks for streaming
     */
    fun synthesizeStream(text: String): Iterator<ByteArray> {
        // Split text into sentences or chunks
        val sentences = text.split(Regex("[.!?]\\s*"))
            .filter { it.isNotBlank() }
        
        return sentences.map { sentence ->
            synthesize(sentence.trim() + ".")
        }.iterator()
    }
    
    /**
     * Unload model and free resources
     */
    fun unload() {
        ortSession?.close()
        ortSession = null
        isModelLoaded = false
    }
    
    /**
     * Check if model is loaded
     */
    fun isLoaded(): Boolean = isModelLoaded
    
    /**
     * Get model information
     */
    fun getModelInfo(): Map<String, Any> {
        return mapOf(
            "loaded" to isModelLoaded,
            "sampleRate" to SAMPLE_RATE,
            "hopLength" to HOP_LENGTH
        )
    }
}

File: ios/Runner/KokoroNative.swift
import Foundation
import Accelerate

class KokoroNative {
    
    private var isModelLoaded = false
    private var modelPath: String?
    
    // ONNX Runtime session
    // private var ortSession: ORTSession?
    
    private let SAMPLE_RATE = 16000
    private let HOP_LENGTH = 256
    
    func loadModel(modelPath: String) -> Bool {
        guard FileManager.default.fileExists(atPath: modelPath) else {
            print("Model file not found: \(modelPath)")
            return false
        }
        
        do {
            // Initialize ONNX Runtime session
            // self.ortSession = try ORTSession(path: modelPath)
            
            self.modelPath = modelPath
            self.isModelLoaded = true
            
            print("TTS model loaded successfully")
            return true
        } catch {
            print("Failed to load TTS model: \(error)")
            return false
        }
    }
    
    func synthesize(text: String, speed: Float = 1.0, voice: String = "female") throws -> Data {
        guard isModelLoaded else {
            throw NSError(domain: "KokoroNative", code: 1, userInfo: [NSLocalizedDescriptionKey: "Model not loaded"])
        }
        
        // Prepare text
        let preparedText = prepareText(text)
        
        // Convert text to token IDs
        let tokenIds = textToTokens(preparedText)
        
        // Get speaker embedding
        let speakerEmbedding = getSpeakerEmbedding(voice: voice)
        
        // Run ONNX inference
        // let melOutput = try runInference(tokenIds: tokenIds, speakerEmbedding: speakerEmbedding)
        
        // Convert mel-spectrogram to waveform
        // let audioFloats = melToWaveform(mel: melOutput, speed: speed)
        
        // Convert to PCM data
        // let pcmData = floatArrayToPCM(audioFloats)
        
        // Placeholder: return silence
        let duration = 2.0 // seconds
        let sampleCount = Int(Double(SAMPLE_RATE) * duration)
        let silence = [Int16](repeating: 0, count: sampleCount)
        let pcmData = Data(bytes: silence, count: silence.count * 2)
        
        return pcmData
    }
    
    private func prepareText(_ text: String) -> String {
        return text
            .trimmingCharacters(in: .whitespacesAndNewlines)
            .replacingOccurrences(of: "\\s+", with: " ", options: .regularExpression)
            .lowercased()
    }
    
    private func textToTokens(_ text: String) -> [Int64] {
        // TODO: Implement proper text-to-phoneme conversion
        let vocab = loadVocabulary()
        
        var tokens: [Int64] = []
        for char in text {
            let token = vocab[String(char)] ?? 0
            tokens.append(token)
        }
        
        return tokens
    }
    
    private func loadVocabulary() -> [String: Int64] {
        // TODO: Load actual vocabulary from tokenizer config
        return [:]
    }
    
    private func getSpeakerEmbedding(voice: String) -> [Float] {
        // Return pre-computed speaker embeddings
        switch voice {
        case "male":
            return [Float](repeating: 0.5, count: 256)
        default: // female
            return [Float](repeating: 0.0, count: 256)
        }
    }
    
    private func runInference(tokenIds: [Int64], speakerEmbedding: [Float]) throws -> [[Float]] {
        // Create ONNX tensors and run inference
        // let inputTensor = try ORTTensor(...)
        // let outputs = try ortSession?.run(withInputs: [...])
        
        // Return mel-spectrogram
        return []
    }
    
    private func melToWaveform(mel: [[Float]], speed: Float) -> [Float] {
        // Convert mel-spectrogram to audio waveform
        // Using Griffin-Lim or neural vocoder
        
        let numFrames = mel.count
        let audioLength = Int(Float(numFrames * HOP_LENGTH) / speed)
        
        return [Float](repeating: 0, count: audioLength)
    }
    
    private func floatArrayToPCM(_ audioFloats: [Float]) -> Data {
        var pcmData = Data(count: audioFloats.count * 2)
        
        pcmData.withUnsafeMutableBytes { (bytes: UnsafeMutableRawBufferPointer) in
            let int16Buffer = bytes.bindMemory(to: Int16.self)
            
            for i in 0..<audioFloats.count {
                // Clamp to [-1, 1] and convert to 16-bit PCM
                let clamped = min(max(audioFloats[i], -1.0), 1.0)
                int16Buffer[i] = Int16(clamped * 32767)
            }
        }
        
        return pcmData
    }
    
    func synthesizeStream(text: String) -> [Data] {
        // Split text into sentences
        let sentences = text.components(separatedBy: CharacterSet(charactersIn: ".!?"))
            .filter { !$0.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty }
        
        var chunks: [Data] = []
        for sentence in sentences {
            do {
                let audioData = try synthesize(text: sentence.trimmingCharacters(in: .whitespaces) + ".", speed: 1.0, voice: "female")
                chunks.append(audioData)
            } catch {
                print("Failed to synthesize sentence: \(error)")
            }
        }
        
        return chunks
    }
    
    func unload() {
        // ortSession = nil
        isModelLoaded = false
        modelPath = nil
        print("TTS model unloaded")
    }
    
    func isLoaded() -> Bool {
        return isModelLoaded
    }
    
    func getModelInfo() -> [String: Any] {
        return [
            "loaded": isModelLoaded,
            "sampleRate": SAMPLE_RATE,
            "hopLength": HOP_LENGTH
        ]
    }
}